File: services/client/src/client/protocol/deserializer.go
package protocol

import (
	"encoding/binary"

	"github.com/7574-sistemas-distribuidos/tp-nivelador/src/client/lottery"
)

func deserializeString(data []byte, length uint16) string {
	return string(data[:length])
}

func deserializeInt(data []byte) int {
	value := BytesToInt(data[:INT_SIZE])
	return value
}

func DeserializeBet(data []byte) (*lottery.Bet, bool) {
	bet := &lottery.Bet{}
	offset := uint32(0)

	for offset < uint32(len(data)) {
		fieldType := FieldType(data[offset])
		if fieldType == EndOfTransmission {
			print("End of transmission reached - offset: ", offset, " total length: ", len(data), "\n")
			return bet, true
		}
		offset++

		length := binary.BigEndian.Uint16(data[offset : offset+LENGTH_SIZE])
		offset += LENGTH_SIZE

		switch fieldType {
		case FieldTypeName:
			value := deserializeString(data[offset:], length)
			bet.FirstName = value
		case FieldTypeLastName:
			value := deserializeString(data[offset:], length)
			bet.LastName = value
		case FieldTypeBirthDate:
			value := deserializeString(data[offset:], length)
			bet.BirthDate = value
		case FieldTypeDocument:
			value := deserializeInt(data[offset:])
			bet.Document = int32(value)
		case FieldTypeLotteryNumber:
			value := deserializeInt(data[offset:])
			bet.LotteryNumber = int32(value)
		case FieldTypeAgencyId:
			value := deserializeInt(data[offset:])
			bet.AgencyId = uint32(value)
		default:
			println("Error: unknown field type")
			return nil, false
		}

		offset += uint32(length)
	}

	return bet, false
}

func DeserializeBatch(data []byte) ([]lottery.Bet, bool) {
	bets := []lottery.Bet{}
	offset := uint32(0)

	for offset < uint32(len(data)) {
		if offset+TOTAL_LENGTH_SIZE > uint32(len(data)) {
			println("Error: not enough data to read bet length")
			break
		}

		betLength := binary.BigEndian.Uint32(data[offset : offset+TOTAL_LENGTH_SIZE])

		if betLength == 0 || offset+betLength > uint32(len(data)) {
			println("Error: invalid bet length or buffer overflow: ", betLength, "offset:", offset, "total length:", len(data))
			break
		}

		offset += TOTAL_LENGTH_SIZE
		betData := data[offset : offset+betLength]

		bet, eot := DeserializeBet(betData)
		if bet == nil {
			// TODO: Handle error case, maybe log it or return an error
			println("Error deserializing bet data")
			return nil, false
		}

		if eot {
			return bets, true
		}
		bets = append(bets, *bet)

		offset += betLength
	}

	return bets, false
}

File: services/server/protocol/deserializer.py
from protocol.utils import (
    END_OF_TRANSMISSION,
    FIELD_TYPE_NAME,
    FIELD_TYPE_LAST_NAME,
    FIELD_TYPE_DOCUMENT,
    FIELD_TYPE_BIRTH_DATE,
    FIELD_TYPE_LOTTERY_NUMBER,
    FIELD_TYPE_AGENCY_ID,
    TOTAL_LENGTH_SIZE,
    TYPE_SIZE,
    LENGTH_SIZE,
    INT_SIZE,
)
from lottery import Bet

def deserialize_int(data: bytes) -> int:
    """Returns the integer value from the given bytes."""
    return int.from_bytes(data[:INT_SIZE], byteorder="big")

def deserialize_string(data: bytes, length: int) -> str:
    """Returns the string value from the given bytes."""
    return data[:length].decode("utf-8")

def deserialize_bet(data: bytes) -> Bet:
    """Deserializes a Bet object from the given bytes. 
    If the FieldType is END_OF_TRANSMISSION, it returns None."""
    bet = {}
    offset = 0

    while offset < len(data):
        field_type = data[offset]
        if field_type == END_OF_TRANSMISSION:
            print("End of transmission reached while deserializing bet.")
            return None

        field_length = int.from_bytes(data[offset + TYPE_SIZE:offset + TYPE_SIZE + LENGTH_SIZE], byteorder="big")
        offset += TYPE_SIZE + LENGTH_SIZE

        if field_type == FIELD_TYPE_NAME:
            bet["first_name"] = deserialize_string(data[offset:], field_length)
        elif field_type == FIELD_TYPE_LAST_NAME:
            bet["last_name"] = deserialize_string(data[offset:], field_length)
        elif field_type == FIELD_TYPE_DOCUMENT:
            bet["document"] = deserialize_int(data[offset:])
        elif field_type == FIELD_TYPE_BIRTH_DATE:
            bet["birthdate"] = deserialize_string(data[offset:], field_length)
        elif field_type == FIELD_TYPE_LOTTERY_NUMBER:
            bet["number"] = deserialize_int(data[offset:])
        elif field_type == FIELD_TYPE_AGENCY_ID:
            bet["agency_id"] = deserialize_int(data[offset:])

        offset += field_length

    agency_id = bet.get("agency_id")
    if agency_id is not None:
        agency_id = int(agency_id)

    return Bet(
        agency_id=agency_id,
        first_name=bet.get("first_name"),
        last_name=bet.get("last_name"),
        document=bet.get("document"),
        birthdate=bet.get("birthdate"),
        number=bet.get("number"),
    )

def deserialize_batch(data: bytes) -> (list[Bet], bool):
    """Deserializes a batch of Bet objects from the given bytes. 
    Returns a tuple containing the list of bets and a boolean indicating
    if End of Transmission was reached during deserialization."""
    bets = []
    offset = 0

    while offset < len(data):
        total_length = int.from_bytes(data[offset:offset + TOTAL_LENGTH_SIZE], byteorder="big")
        offset += TOTAL_LENGTH_SIZE

        bet_data = data[offset:offset + total_length]
        bet = deserialize_bet(bet_data)
        if bet is not None:
            bets.append(bet)
        else:
            return bets, True

        offset += total_length

    return bets, False
